src/validation/input-params.ts

```diff
@@ -1,28 +1,478 @@
+ * Type for the overrides object that is either hardcoded in the adapter,
+ * or present in the incoming request body
+ */
 export type Override = Map<string, Map<string, string>>
 
-type ParameterType = 'bigint' | 'boolean' | 'array' | 'number' | 'object' | 'string'
+/**
+ * Possible strings that correspond to types of primitive values
+ * that can be specified for an input parameter
+ */
+type PrimitiveParameterTypeString = 'boolean' | 'number' | 'string'
 
-export type InputParameter = {
-  aliases?: readonly string[]
-  description?: string
-  type?: ParameterType
+/**
+ * Possible types for each of the possible type strings
+ */
+type PrimitiveParameterType = TypeFromTypeString<PrimitiveParameterTypeString>
+
+/**
+ * All possible types for an input parameter (either a string, or a nested param definition)
+ */
+type ParameterType = PrimitiveParameterTypeString | InputParametersDefinition
+
+/**
+ * Given a type parameter T that is a parameter type (string or nested definition),
+ * equals to the corresponding type for that type string or definition
+ */
+type TypeFromTypeString<T extends ParameterType> = T extends 'string'
+  ? string
+  : T extends 'number'
+  ? number
+  : T extends 'boolean'
+  ? boolean
+  : T extends InputParametersDefinition
+  ? TypeFromDefinition<T>
+  : never
+
+/**
+ * Util type to be able to discern exactly if a generic is exactly unknown
+ * (because everything extends unknown, but unknown only extends itself)
+ */
+type IsUnknown<T> = unknown extends T ? true : false
+
+/**
+ * Given a concrete InputParameter P, results in the type for that definition taking into account
+ * the required property of the definition (i.e. makes it optional or not)
+ */
+type ShouldBeUndefinable<
+  P extends InputParameter,
+  T = TypeFromTypeString<P['type']>,
+> = P['required'] extends true ? T : IsUnknown<P['default']> extends true ? T | undefined : T
+
+/**
+ * Just an alias to make types below more readable
+ */
+type NonArrayInputType<P extends InputParameter> = ShouldBeUndefinable<P>
+
+/**
+ * Given an InputParameter P, results in the corresponding type for that definition,
+ * accounting for configurations like required, array, defaults, etc.
+ */
+type TypeFromParameter<
+  P extends InputParameter,
+  T = TypeFromTypeString<P['type']>,
+> = P['array'] extends true ? T[] : NonArrayInputType<P>
+
+/**
+ * Constraint for an InputParameter to make sure options are only specified for parameters of type string.
+ * All these constraints, for them to work properly, should represent the entire closed realm of possibilities.
+ * That is, there should be no possible input parameter that the constraint doesn't accept one way or the other.
+ */
+type InputParameterOptionConstraints =
+  | {
+      type: 'string'
+      options?: string[] | readonly string[]
+    }
+  | {
+      type: 'number'
+      options?: number[] | readonly number[]
+    }
+  | {
+      type: Exclude<ParameterType, 'string'>
+      options?: never
+    }
```

Comment:
Wouldn't this need to be this now?
```suggestion
  | {
      type: Exclude<ParameterType, 'string' | 'number'>
      options?: never
    }
```

Reply:
Fixed